File: 169-majority-element/169-majority-element.scala
object Solution {
    def majorityElement(nums: Array[Int]): Int = {
        var cur: Int = 0
        var cnt: Int = 0
        
        nums.foreach { n => 
            if (cnt == 0) {
                cur = n
                cnt = 1
            } else {
                cnt = if (cur == n) cnt + 1 else cnt - 1
            }
        }
        
        cur
    }
}

File: 169-majority-element/169-majority-element.java
class Solution {
    public int majorityElement(int[] nums) {
        int cnt = 0;
        Integer current = null;
        
        for(int i = 0; i < nums.length; i++) {
            if (current == null) {
                current = nums[i];
                cnt = 1;
                continue;
            }
            
            if (nums[i] == current) {
                cnt += 1;
                continue;
            }
            
            cnt -= 1;
            if (cnt == 0) {
                current = null;
            }
        }
        
        return current;
    }
}